xonsh/cli_utils.py

```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
+        from pygments.lexers.markup import RstLexer
+
+        return highlight(text, RstLexer(), self.formatter)
+
+    def _format_text(self, text):
+        text = super()._format_text(text)
+        if text:
+            text = self.markup_rst(text)
+        return text
+
+    def _format_usage(self, usage, actions, groups, prefix):
+        from pygments.token import Name, Generic
+
+        text = super()._format_usage(usage, actions, groups, prefix)
+        parts = text.split(self._prog, maxsplit=1)
+        if len(parts) == 2 and all(parts):
+            text = self.colorize(
+                (Generic.Heading, parts[0]),
+                (Name.Function, self._prog),
+                (Name.Attribute, parts[1]),  # from _format_actions_usage
+            )
+        return text
+
+    def _format_action_invocation(self, action):
+        from pygments.token import Name
+
+        text = super()._format_action_invocation(action)
+        return self.colorize((Name.Attribute, text))
+
+
+def get_argparse_formatter_class():
+    from xonsh.built_ins import XSH
+    from xonsh.platform import HAS_PYGMENTS
+
+    if (
+        hasattr(sys, "stderr")
+        and sys.stderr.isatty()
+        and XSH.env.get("XONSH_INTERACTIVE")
+        and HAS_PYGMENTS
+    ):
+        return RstHelpFormatter
+    return ap.RawTextHelpFormatter
+
+
+class ArgParser(ap.ArgumentParser):
+    """Sub-class of ArgumentParser with special methods to nest commands"""
+
+    def __init__(self, **kwargs):
+        if "formatter_class" not in kwargs:
+            kwargs["formatter_class"] = get_argparse_formatter_class()
+
+        super().__init__(**kwargs)
+        self.commands = None
+
+    def add_command(
+        self, func: tp.Callable, args: tp.Optional[tp.Iterable[str]] = None, **kwargs
+    ):
+        """
+            create a sub-parser and call this function during dispatch
+
+        Parameters
+        ----------
+        func
+            a type-annotated function that will be used to create ArgumentParser instance.
+            All parameters that start with ``_`` will not be added to parser arguments.
+            Use _stdout, _stack ... to receive them from callable-alias/commands.
+            Use _parser to get the generated parser instance.
+            Use _args to get what is passed from sys.argv
+            Use _parsed to get result of ``parser.parse_args``
+        args
+            if given only add these arguments to the parser.
+            Otherwise all parameters to the function without `_` prefixed
+            in their name gets added to the parser.
+        kwargs
+            passed to ``subparser.add_parser`` call
+
+        Returns
+        -------
+            result from ``subparser.add_parser``
+        """
+        if not self.commands:
+            self.commands = self.add_subparsers(title="commands", dest="command")
+
+        doc = get_doc(func)
+        kwargs.setdefault("description", doc)
```

Comment:
Also here set description to be `help`'s first line (bc the newline doesn't render well in the completion's description)

Reply:
this is the content shown in the help as well. I think the responsibility of what to show in the completion-widget should be left to widget itself.

Reply:
Sure, but if it doesn't define one, why include the newlines that we know aren't being displayed correctly currently on ptk?

Reply:
Here we don't have to assume it is ptk. Perhaps the ptk completer class can update this completions?